### PR TITLE
Improved test coverage, removed access token method const

### PR DESCRIPTION
**Pull request by rtheunissen**
This pull request achieves the following:

 - Improved test coverage
 - Ability to test arbitrary access token request methods

### CODE REVIEW

File: src/Provider/AbstractProvider.php [hunk 1]
```diff
@@ -88,6 +84,23 @@
...
+     * @return array
+     */
+    abstract protected function getDefaultScopes();
+
+    /**
+     * Returns the method to use when requesting an access token.
+     *
+     * @return string HTTP method
+     */
+    abstract protected function getAccessTokenMethod();
```
Comment (rtheunissen) [inline src/Provider/AbstractProvider.php]: This replaces the CONST. It's still only internal but is more flexible, as can be seen when testing various access token methods,

File: src/Provider/AbstractProvider.php [hunk 2]
```diff
@@ -271,7 +274,7 @@ public function getAuthorizationUrl(array $options = [])
 
         $options += [
             'response_type'   => 'code',
-            'approval_prompt' => 'auto',
+            'approval_prompt' => 'auto'
```
Comment (rtheunissen) [inline src/Provider/AbstractProvider.php]: Personally I always use a trailing comma but it affected test coverage :p. Happy to change it back to preserve consistency.

File: src/Provider/AbstractProvider.php [hunk 3]
```diff
@@ -320,26 +324,28 @@ public function getAccessToken($grant = 'authorization_code', array $params = []
...
+        $method = strtoupper($this->getAccessTokenMethod());
+        $options = [];
 
-        switch (strtoupper(static::ACCESS_TOKEN_METHOD)) {
+        switch ($method) {
             case 'GET':
-                // @codeCoverageIgnoreStart
                 // No providers included with this library use get but 3rd parties may
-                $urlAccessToken .= (strpos($urlAccessToken, '?') ? '&' : '?') . $requestParams;
+                $url .= (parse_url($url, PHP_URL_QUERY) ? '&' : '?') . $requestParams;
```
Comment (rtheunissen) [inline src/Provider/AbstractProvider.php]: This is the nicest way that I could come up with to append the request parameters. Feels a bit safer than checking if '?' is in there somewhere.

File: src/Provider/AbstractProvider.php [hunk 4]
```diff
@@ -320,26 +324,28 @@ public function getAccessToken($grant = 'authorization_code', array $params = []
...
                 break;
-                // @codeCoverageIgnoreEnd
             case 'POST':
                 $options['body'] = $requestParams;
                 break;
-            // @codeCoverageIgnoreStart
             default:
-                throw new \InvalidArgumentException('Neither GET nor POST is specified for request');
-            // @codeCoverageIgnoreEnd
+                throw new InvalidArgumentException(
```
Comment (rtheunissen) [inline src/Provider/AbstractProvider.php]: Removed explicit GET and POST in the message, showing instead what method was attempted. This is mostly to keep things consistent if we were to ever add another method.

File: test/src/Provider/Fake.php [hunk 5]
```diff
@@ -31,6 +33,16 @@ protected function getDefaultScopes()
         return ['test'];
     }
 
+    public function setAccessTokenMethod($method)
+    {
+        $this->accessTokenMethod = $method;
+    }
+
+    public function getAccessTokenMethod()
```
Comment (rtheunissen) [inline test/src/Provider/Fake.php]: Does changing a protected parent method to public when overriding cause issues on some versions?

File: src/Tool/RequestFactory.php [hunk 6]
```diff
@@ -20,36 +29,49 @@ public function getRequest(
...
-     * Get a request using a simplified array of options.
+     * Parses simplified options.
+     *
+     * @param array $options Simplified options.
+     *
+     * @return array Extended options for use with getRequest.
+     */
+    protected function parseOptions(array $options)
+    {
+        // Should match default values for getRequest
```
Comment (rtheunissen) [inline src/Tool/RequestFactory.php]: Not as DRY as @shadowhand's solution using Reflection, but it doesn't use Reflection and the defaults aren't cryptic.

File: src/Tool/RequestFactory.php [hunk 7]
```diff
@@ -20,36 +29,49 @@ public function getRequest(
         $uri,
         array $headers = [],
         $body = null,
-        $protocolVersion = '1.1'
+        $version = self::DEFAULT_PROTOCOL_VERSION
```
Comment (shadowhand) [inline src/Tool/RequestFactory.php]: What's the benefit of doing this? Since it is using `self` it can't be overloaded, making it effectively the same as a string default.

File: src/Tool/RequestFactory.php [hunk 8]
```diff
@@ -20,36 +29,49 @@ public function getRequest(
...
-
-        return $func->invokeArgs($this, array_values($args));
+        $options = $this->parseOptions($options);
+
+        return $this->getRequest(
+            $method,
+            $uri,
+            $options['headers'],
+            $options['body'],
+            $options['version']
```
Comment (shadowhand) [inline src/Tool/RequestFactory.php]: Yeah, probably makes sense to do this.

File: test/src/Provider/AbstractProviderTest.php [hunk 9]
```diff
@@ -267,65 +267,6 @@ public function testRandomGeneratorCreatesRandomState()
         $this->assertRegExp('/^[a-zA-Z0-9\/+]{32}$/', $qs['state']);
     }
 
-    public function testGetAccessToken()
```
Comment (shadowhand) [inline test/src/Provider/AbstractProviderTest.php]: Why is this deleted?

File: test/src/Provider/AbstractProviderTest.php [hunk 10]
```diff
@@ -472,4 +414,85 @@ public function testAuthenticatedRequestAndResponse()
...
+
+        $this->assertSame($result, $token);
+        $this->assertSame($raw_response['uid'], $token->getUid());
+        $this->assertSame($raw_response['access_token'], $token->getToken());
+        $this->assertSame($raw_response['expires'], $token->getExpires());
+    }
+
+    public function testGetAccessTokenMethods()
+    {
+        $this->accessTokenTest('GET');
```
Comment (shadowhand) [inline test/src/Provider/AbstractProviderTest.php]: Should be using a `@dataProvider` to do this kind of thing.

File: test/src/Provider/Fake.php (hunk 5 above)
Comment (shadowhand) [inline test/src/Provider/Fake.php]: I don't see why this is necessary?

File: src/Tool/RequestFactory.php (hunk 7 above)
Comment (rtheunissen) [inline src/Tool/RequestFactory.php]: True re: `self`, but you can't use `static` in a default parameter value. I don't know, wanted to give the option of an alternative default protocol version but an extending class can just handle that internally. I'll remove. :+1:

File: test/src/Provider/AbstractProviderTest.php (hunk 10 above)
Comment (rtheunissen) [inline test/src/Provider/AbstractProviderTest.php]: I don't know what that is, sorry. The idea was just to not create two almost-identical test methods.

File: test/src/Provider/Fake.php (hunk 5 above)
Comment (rtheunissen) [inline test/src/Provider/Fake.php]: For testing.

Comment (rtheunissen) [inline test/src/Provider/Fake.php]: Oh, could make it a public property instead? Is that what you meant?

File: test/src/Provider/AbstractProviderTest.php (hunk 10 above)
Comment (shadowhand) [inline test/src/Provider/AbstractProviderTest.php]:
```php
/**
 * @dataProvider getAccessTokenMethodProvider
 */
public function testGetAccessTokenMethods($method) { ... }

public function getAccessTokenMethodProvider()
{
    return [
        ['GET'],
        ['POST'],
    ];
}
```

Comment (rtheunissen) [inline test/src/Provider/AbstractProviderTest.php]: And then `foreach` in the test?

Comment (rtheunissen) [inline test/src/Provider/AbstractProviderTest.php]: Cool :)

File: src/Provider/AbstractProvider.php (hunk 4 above)
Comment (shadowhand) [inline src/Provider/AbstractProvider.php]: I would greatly prefer to see this be done using `sprintf` because: a) it's safer, and b) easier to read:

```php
throw new InvalidArgumentException(sprintf(
    'Unsupported access token request method: "%s"',
    $method
));
```

File: test/src/Provider/AbstractProviderTest.php (hunk 10 above)
Comment (shadowhand) [inline test/src/Provider/AbstractProviderTest.php]: https://phpunit.de/manual/current/en/writing-tests-for-phpunit.html#writing-tests-for-phpunit.data-providers for complete reference.

File: test/src/Provider/Fake.php (hunk 5 above)
Comment (shadowhand) [inline test/src/Provider/Fake.php]: No, I just don't see why you would need to read it this way. You can use `assertAttributeEquals` to read protected properties, or just read it through mocking.

File: src/Provider/AbstractProvider.php (hunk 4 above)
Comment (rtheunissen) [inline src/Provider/AbstractProvider.php]: How is it safer than interpolation? Seeing as the `$method` will be __toString()'d anyway. I don't see how it's easier to read either.. especially as there is only one param. If string interpolation wasn't a thing and we were concatenating then yes, for sure.

```
throw new InvalidArgumentException(
    "Unsupported access token request method: '$method'",
);
```

vs.

```
throw new InvalidArgumentException(sprintf(
    'Unsupported access token request method: "%s"',
    $method
));
```

Comment (rtheunissen) [inline src/Provider/AbstractProvider.php]: Maybe @ramsey could cast a vote. :p

Comment (shadowhand) [inline src/Provider/AbstractProvider.php]: I have an aversion to string interpolation in general, I find it really confusing to read the strings. And if in the future, $method wasn't a string, but a call, it would be trivial to change.

Comment (shadowhand) [inline src/Provider/AbstractProvider.php]: For example https://github.com/thephpleague/oauth2-client/blame/0c3f84b606e155bd780e00dff933f3989b337ece/src/Grant/GrantFactory.php#L81-L84

File: test/src/Provider/Fake.php (hunk 5 above)
Comment (rtheunissen) [inline test/src/Provider/Fake.php]: `assertAttributeEquals` wouldn't help, because we're not asserting, we're using.
Mocking, sure, but you would have to mock the provider whenever you want to use `getAccessTokenMethod`. `Fake` is a test object anyway, I don't see why that's an issue. Feel free to mock instead.

File: src/Provider/AbstractProvider.php (hunk 4 above)
Comment (rtheunissen) [inline src/Provider/AbstractProvider.php]: I'm still in the other camp. In that case I would use:
```
public function checkGrant($class)
{
  if ( ! $this->isGrant($class)) {
      $class = is_object($class) ? get_class($class) : $class;
      throw new InvalidGrantException(sprintf(
          "Grant '$class' must implement GrantInterface"
      ));
  }
}
```

This is just a style conflict. Either we define a policy or nominate someone to make the call.

File: src/Provider/AbstractProvider.php [hunk 11]
```diff
@@ -82,6 +84,23 @@
...
+     * @return array
+     */
+    abstract protected function getDefaultScopes();
+
+    /**
+     * Returns the method to use when requesting an access token.
+     *
+     * @return string HTTP method
+     */
+    abstract protected function getAccessTokenMethod();
```
Comment (shadowhand) [inline src/Provider/AbstractProvider.php]: This shouldn't be a requirement for implementing providers, it should default to `post`.

File: src/Provider/AbstractProvider.php (hunk 4 above)
Comment (rtheunissen) [inline src/Provider/AbstractProvider.php]: If we had Python's `'{name}'.format(name=value)` then I'd be all for it. I just don't like matching parameter order / having to adjust parameter order when I want to change the message later.

**Comment by shadowhand** [comment]
Please rebase, I just brought my branch in sync with 1.0.

File: src/Provider/AbstractProvider.php (hunk 11 above)
Comment (rtheunissen) [inline src/Provider/AbstractProvider.php]: Can we make it protected, but not abstract?

Comment (shadowhand) [inline src/Provider/AbstractProvider.php]: Yes that's exactly what you want.

File: src/Provider/AbstractProvider.php [hunk 12]
```diff
@@ -82,6 +84,23 @@
...
 
     /**
+     * Get the default scopes used by this provider.
+     *
+     * This should not be a complete list of all scopes, but the minimum
+     * required for the provider user interface!
+     *
+     * @return array
+     */
+    abstract protected function getDefaultScopes();
```
Comment (shadowhand) [inline src/Provider/AbstractProvider.php]: Why are these two methods moved above `__construct` ?

**Comment by rtheunissen** [comment]
To do that do I just set your branch as upstream, and run `git rebase`?

File: src/Provider/AbstractProvider.php (hunk 11 above)
Comment (rtheunissen) [inline src/Provider/AbstractProvider.php]: Cool thanks.

File: src/Provider/AbstractProvider.php (hunk 12 above)
Comment (rtheunissen) [inline src/Provider/AbstractProvider.php]: It's nice to be able to see what abstract methods to implement. Maybe below `__construct` makes more sense. I'm impartial though, your call.

Comment (shadowhand) [inline src/Provider/AbstractProvider.php]: I don't feel strongly about it either way, but moving methods arbitrarily is outside the scope of this PR.

Comment (rtheunissen) [inline src/Provider/AbstractProvider.php]: Good call.

**Comment by shadowhand** [comment]
```bash
git fetch shadowhand
git rebase shadowhand/feature/guzzle-6-http
```

**Comment by rtheunissen** [comment]
Okay, hope I did that right.

**Comment by shadowhand** [comment]
Well, it ended up as a merge, but I can work with that.

File: src/Provider/AbstractProvider.php (hunk 2 above)
Comment (shadowhand) [inline src/Provider/AbstractProvider.php]: I don't see how a comma would affect test coverage.

**Comment by rtheunissen** [comment]
Re: rebase, I haven't done it before so I'm not surprised that it didn't go as planned. :/ Next time.

**Comment by shadowhand** [comment]
Right, you always have to use a force push after a rebase, because it rewrites history.

**Comment by rtheunissen** [comment]
> Right, you always have to use a force push after a rebase, because it rewrites history.

Thanks :dart:
